src/osd/OSD.cc

```diff
@@ -3888,6 +3888,10 @@ int OSD::init()
   // load up pgs (as they previously existed)
   load_pgs();
 
+  // fix omap
+  store->fix_omap();
```

Comment:
From my experience doing potentially long-lasting format changes during or immediately after an upgrade is not the best practice. And scanning through all the OMAPs and unconditionally converting them could indeed take a while. We had similar issues before when converting to per-pool/per-pg OMAP format.  
IMO it would be better to have this new format update performed in the same way. Which is introducing additional flag to onode to indicate new vs. old format usage and doing format update on store's repair. Possibly coupled with cluster wide alert that legacy format is being used if old format is detected. 
Please see https://github.com/ceph/ceph/pull/38651 as a referral how we implemented such stuff last time.